### PR TITLE
8338479: GenShen: Detemplatize ShenandoahScanRemembered

### PR DESCRIPTION
**Pull request by ysramakrishna**
De-templatize ShenandoahScanRemembered and ShenandoahCardCluster, based on Roman's review feedback at https://github.com/openjdk/jdk/pull/20395/files/46638cdef26fd25c290e6f63035ff5d2f1dc3bdb#r1718773835.

/issue JDK-8338479

**Testing:**

- [x] GHA
- [x] jtreg:tier1
- [x] codepipeline
- [x] SPECjbb

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8338479](https://bugs.openjdk.org/browse/JDK-8338479): GenShen: Detemplatize ShenandoahScanRemembered (**Task** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer) ⚠️ Review applies to [b97b7a14](https://git.openjdk.org/shenandoah/pull/472/files/b97b7a14a4b06e080623cdf2e32b0310ca603ffd)
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/472/head:pull/472` \
`$ git checkout pull/472`

Update a local copy of the PR: \
`$ git checkout pull/472` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/472/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 472`

View PR using the GUI difftool: \
`$ git pr show -t 472`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/472.diff">https://git.openjdk.org/shenandoah/pull/472.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/472#issuecomment-2293669266)

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp:1]: These method definitions have been moved verbatim to the .cpp file.

File: src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.cpp:1]: These method definitions have moved verbatim from the .inline.hpp.

**Review by kdnilsen — APPROVED** [review]
Thanks for quick turnaround on this.  I assume we'll complete all tests before integrating.

File: src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp [hunk 1]
```diff
@@ -225,19 +223,19 @@ class ShenandoahDirectCardMarkRememberedSet: public CHeapObj<mtGC> {
   size_t total_cards() const;
   size_t card_index_for_addr(HeapWord *p) const;
   HeapWord *addr_for_card_index(size_t card_index) const;
-  inline const CardValue* get_card_table_byte_map(bool write_table) const;
```
Comment (earthling-amzn) [inline src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp:228]: Should these smaller, non-templated functions remain inline?

Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp:228]: ok, let me revert the specifier back.

**Comment by ysramakrishna** [comment]
Running specjbb perf numbers and codepipeline again after syncing with latest changes in master. Will push later today once all is clear and no perf differences (don't expect any). Left the inline specifier intact on those methods as requested by William.

Thanks for your reviews!

File: src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp [hunk 2]
```diff
@@ -1018,14 +996,13 @@ class ShenandoahRegionChunkIterator : public StackObj {
 
   // Fills in assignment with next chunk of work and returns true iff there is more work.
   // Otherwise, returns false.  This is multi-thread-safe.
-  inline bool next(struct ShenandoahRegionChunk *assignment);
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp:1021]: Missed these; will restore these as well.

Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp]: restored.

**Review by earthling-amzn — APPROVED** [review]

**Comment by ysramakrishna** [comment]
Thanks for your reviews!

/integrate

**Pull request closed without merging**